### PR TITLE
workbench implementation for OneStepResampling.sh

### PR DESCRIPTION
**Pull request by coalsont**
This cannot be used by the public yet (not easily, anyway), as the -volume-resample command is not released, nor is the control of volume output datatype (added to resolve a considerable file size change between applywarp and wb_command output), so it is disabled by default.  Currently, enabling it requires editing the GenericfMRIVolume pipeline.

Once workbench has a new release, it can be made the default, or the fslsplit/loop/fslmerge code can be removed entirely.

### CODE REVIEW

File: fMRIVolume/scripts/OneStepResampling.sh [hunk 1]
```diff
@@ -22,7 +22,6 @@ Usage: ${script_name} [options]
   --t1=<input T1w restored image>
   --fmriresout=<output resolution for images, typically the fmri resolution>
   --fmrifolder=<fMRI processing folder>
-  --atlasspacedir=<output directory for several resampled images>
```
Comment (glasserm) [inline fMRIVolume/scripts/OneStepResampling.sh:25]: Why was this removed?

Comment (coalsont) [reply fMRIVolume/scripts/OneStepResampling.sh:25]: Because the option doesn't exist.

**Comment by coalsont** [comment]
For reference, the usage of the unreleased -volume-resample:

```$ wb_command -volume-resample 
TRANSFORM AND RESAMPLE A VOLUME FILE
   wb_command -volume-resample
      <volume-in> - volume to resample
      <volume-space> - a volume file in the volume space you want for the
         output
      <method> - the resampling method
      <volume-out> - output - the output volume

      [-affine] - repeatable - add an affine transform
         <affine> - the affine file to use

         [-flirt] - MUST be used if affine is a flirt affine
            <source-volume> - the source volume used when generating the affine
            <target-volume> - the target volume used when generating the affine

      [-affine-series] - repeatable - add an independent affine per-frame
         <affine-series> - text file containing 12 or 16 numbers per line, each
            being a row-major flattened affine

         [-flirt] - MUST be used if affine is a flirt affine
            <source-volume> - the source volume used when generating the affine
            <target-volume> - the target volume used when generating the affine

      [-warp] - repeatable - add a nonlinear warpfield transform
         <warpfield> - the warpfield file

         [-fnirt] - MUST be used if using a fnirt warpfield
            <source-volume> - the source volume used when generating the
               warpfield

      Resample a volume file with an arbitrary list of transformations.  You
      may specify -affine, -warp, and -affine-series multiple times each, and
      they will be used in the order specified.  For instance, for rigid motion
      correction followed by nonlinear atlas registration, specify
      -affine-series first, then -warp.  The recommended methods are CUBIC
      (cubic spline) for most data, and ENCLOSING_VOXEL for label data.  The
      parameter <method> must be one of:

      CUBIC
      ENCLOSING_VOXEL
      TRILINEAR
```

File: fMRIVolume/scripts/OneStepResampling.sh [hunk 2]
```diff
@@ -244,50 +260,93 @@ ${FSLDIR}/bin/imcp ${WD}/${T1wImageFile}.${FinalfMRIResolution} ${fMRIFolder}/${
...
+        if [ $k -eq 0 ] ; then
+            echo "0" >> ${fMRIFolder}/Movement_RelativeRMS.txt
+        else
+            rmsdiff ${MotionMatrixFolder}/${MotionMatrixPrefix}${vnum} $prevmatrix ${ScoutInputgdc} ${ScoutInputgdc}_mask.nii.gz | tail -n 1 >> ${fMRIFolder}/Movement_RelativeRMS.txt
+        fi
+        prevmatrix="${MotionMatrixFolder}/${MotionMatrixPrefix}${vnum}"
+    done
+    
+    #gdc warp space is input to input, affine is input to input, OutputTransform is input to MNI
+    xfmargs=(-warp "$GradientDistortionField.nii.gz" -fnirt "$InputfMRI"
```
Comment (glasserm) [inline fMRIVolume/scripts/OneStepResampling.sh:287]: I prefer ${}, particularly when relying on non text for the parsing into variable/non-variable.  I was confused reading this line.

Comment (coalsont) [reply fMRIVolume/scripts/OneStepResampling.sh:287]: Another option would be `"$GradientDistortionField".nii.gz`.  Quotes do something useful, braces generally do not.

Comment (glasserm) [reply fMRIVolume/scripts/OneStepResampling.sh:287]: Yes that would be clearer too.

File: fMRIVolume/scripts/MotionCorrection.sh [hunk 3]
```diff
@@ -181,7 +181,7 @@ function DeriveBackwards {
   # TCS becomes an array of the values from column $i in $in (derived from Var)
   TCS=($Var)
   # random is a random file name for temporary output
-  random=$RANDOM
+  random="$(mktemp --tmpdir MotionCorrectionRandom_XXXXXX.txt)"
```
Comment (mharms) [inline fMRIVolume/scripts/MotionCorrection.sh:184]: `--tmpdir` isn't a supported option on a standard MacOS installation.

Comment (mharms) [reply fMRIVolume/scripts/MotionCorrection.sh:184]: Also, if going this route, should we also make use of the `tempfiles.shlib` in this script as well?

File: fMRIVolume/scripts/OneStepResampling.sh [hunk 4]
```diff
@@ -135,6 +135,22 @@ JacobianOut=`getopt1 "--ojacobian" $@`  # "${18}"
 fMRIReferencePath=`getopt1 "--fmrirefpath" $@` # "${19}"
 fMRIReferenceReg=`getopt1 "--fmrirefreg" $@`  # "${20}"
 
+#hidden: toggle for unreleased new resampling command, default off
```
Comment (mharms) [inline fMRIVolume/scripts/OneStepResampling.sh:138]: Maybe add a comment explaining the advantage of the `--wb-resample` approach?

File: fMRIVolume/scripts/OneStepResampling.sh [hunk 5]
```diff
@@ -244,50 +260,93 @@ ${FSLDIR}/bin/imcp ${WD}/${T1wImageFile}.${FinalfMRIResolution} ${fMRIFolder}/${
...
-  ${FSLDIR}/bin/convertwarp --relout --rel --ref=${WD}/${T1wImageFile}.${FinalfMRIResolution} --warp1=${MotionMatrixFolder}/${MotionMatrixPrefix}${vnum}_gdc_warp.nii.gz --warp2=${OutputTransform} --out=${MotionMatrixFolder}/${MotionMatrixPrefix}${vnum}_all_warp.nii.gz
-  # Apply one-step warp, using spline interpolation
-  ${FSLDIR}/bin/applywarp --rel --interp=spline --in=${WD}/prevols/vol${vnum}.nii.gz --warp=${MotionMatrixFolder}/${MotionMatrixPrefix}${vnum}_all_warp.nii.gz --ref=${WD}/${T1wImageFile}.${FinalfMRIResolution} --out=${WD}/postvols/vol${vnum}.nii.gz
 
-  # Generate a mask for keeping track of spatial coverage (use nearest neighbor interpolation here)
-  ${FSLDIR}/bin/fslmaths ${WD}/prevols/vol${vnum}.nii.gz -mul 0 -add 1 ${WD}/prevols/vol${vnum}_mask.nii.gz
-  ${FSLDIR}/bin/applywarp --rel --interp=nn --in=${WD}/prevols/vol${vnum}_mask.nii.gz --warp=${MotionMatrixFolder}/${MotionMatrixPrefix}${vnum}_all_warp.nii.gz --ref=${WD}/${T1wImageFile}.${FinalfMRIResolution} --out=${WD}/postvols/vol${vnum}_mask.nii.gz
+if ((useWbResample))
+then
+    affseries="$(mktemp --tmpdir OneStepResampleAffSeries_XXXXXX.txt)"
```
Comment (mharms) [inline fMRIVolume/scripts/OneStepResampling.sh:266]: Ditto comment about `--tmpdir` on MacOS.

File: fMRIVolume/scripts/OneStepResampling.sh [hunk 6]
```diff
@@ -244,50 +260,93 @@ ${FSLDIR}/bin/imcp ${WD}/${T1wImageFile}.${FinalfMRIResolution} ${fMRIFolder}/${
...
+    #gdc warp space is input to input, affine is input to input, OutputTransform is input to MNI
+    xfmargs=(-warp "$GradientDistortionField.nii.gz" -fnirt "$InputfMRI"
+             -affine-series "$affseries" -flirt "$InputfMRI" "$InputfMRI"
+             -warp "$OutputTransform.nii.gz" -fnirt "$InputfMRI")
+    
+    wb_command -volume-resample "$InputfMRI" "$WD/$T1wImageFile.$FinalfMRIResolution.nii.gz" CUBIC "$OutputfMRI.nii.gz" "${xfmargs[@]}" -nifti-output-datatype INT32
+    
+    #resample all-ones volume series with enclosing voxel to determine FOV coverage
+    #yes, this is the entire length of the timeseries on purpose
+    fovcheck="$(mktemp --tmpdir OneStepResampleFovCheck_XXXXXX.nii.gz)"
```
Comment (mharms) [inline fMRIVolume/scripts/OneStepResampling.sh:295]: Ditto `--tmpdir`

**Comment by glasserm** [comment]
${OutputfMRI} is equivalent for processing.  Perhaps Tim can check the other two quickly.

File: fMRIVolume/scripts/MotionCorrection.sh (hunk 3 above)
Comment (coalsont) [reply fMRIVolume/scripts/MotionCorrection.sh:184]: Hmm...I had already used `--tmpdir` elsewhere, that is unfortunate...

`tempfiles.shlib` only contains functions for cleanup, it doesn't actually generate any temporary file names currently.  If we need some compatibility workaround on Mac, then maybe it should go in there.

**Comment by coalsont** [comment]
The mask from the workbench resampling has an extra partial voxel on its FoV (the difference image has an interrupted line of voxels on the visible parts of the original FoV edge).  I was expecting less of a difference than that, but it is true that our enclosing voxel method is half a voxel more generous than interpolation (which happens to also be 1% of a voxel more generous than the naive approach, on purpose).  Trilinear should actually be much closer to the fsl "nn" result for this case of all-ones input, while still being much faster (mostly by avoiding two thousand convertwarp calls - these could probably be reduced to one thousand by using --midmat in convertwarp).

Movement_AbsoluteRMS.txt and Movement_RelativeRMS.txt are identical.  The _mean files happen to be missing from my testing folder at the moment, but that is because I interrupted the most recent testing run while trying to pin down something else, they should also be identical.

File: fMRIVolume/scripts/MotionCorrection.sh (hunk 3 above)
Comment (coalsont) [reply fMRIVolume/scripts/MotionCorrection.sh:184]: On the old mac system I can easily test on, mktemp doesn't even allow generating a temporary filename with a desired file extension:

```
$ mktemp foo-XXXXXXX.txt
foo-XXXXXXX.txt
$ mktemp -t foo-XXXXXXX.txt
/tmp/foo-XXXXXXX.txt.laTQHK7E
```

This is pretty disappointing.  I guess I can write a function that uses --tmpdir on everything except mac, and fake it on mac, but this means I have to detect mac systems reliably.

Comment (mharms) [reply fMRIVolume/scripts/MotionCorrection.sh:184]: I know `mktemp` may be elegant, but why not just stick with the existing code of writing to a file named from `$RANDOM`?  We are deleting the file anyway.

Comment (coalsont) [reply fMRIVolume/scripts/MotionCorrection.sh:184]: I edited it because in my test run, that folder was not writable.  Using a folder that is always writable for all temporaries is better practice than ever using the current directory, and using $RANDOM as a mechanism for trying to avoid filename collision is bad when there is already something designed for specifically that purpose.

**Comment by coalsont** [comment]
I have pushed edits to resolve these concerns, currently being tested.  However, it appears the change to trilinear interpolation for the FoV coverage mask didn't change as much as I expected, there is still an extra half voxel or so on parts of the edge.  I'm not really sure what to make of that.

I also hadn't realized it, but on one edge (P), the FSL mask has an extra half voxel or so, rather than the workbench result.  Chasing this down could get tricky, how worried about this are we?

Anecdotal timing info, one run of only OneStepResample, 3T REST1_LR:
new wb_command code:
```
real	12m28.293s
user	27m43.616s
sys	0m14.204s
```
old applywarp code:
```
over an hour so far, looks like probably 4 hours total, will update when it finishes
```
I didn't notice anything indicating problems with nfs bandwidth/latency, but the applywarp time should be checked against prior timings of GenericfMRIVolume for sanity.  The applywarp code could be made somewhat faster by using --midmat in convertwarp, but it would still do fslsplit, fslmerge, and one convertwarp per frame.  The different ref spaces should be fine (just use the final one), since each transform matches its input space to the previous output space.

As an aside, the practical advantage of the wb_command -fnirt and -flirt setup (besides identity transforms actually meaning what they say, and being able to transform surface files) is that it can use transforms that do NOT match every step's output grid to the next input grid (remember the applyisoxfm stuff to make different resolutions play nice? that doesn't matter to workbench).  If we made "world" convention versions of transforms that we planned to use in wb_command, then we wouldn't need to use -flirt and -fnirt arguments (or remember what specific spaces they use).

File: fMRIVolume/scripts/OneStepResampling.sh (hunk 4 above)
Comment (coalsont) [reply fMRIVolume/scripts/OneStepResampling.sh:138]: I added a comment under the option parsing lines - I expect when workbench is released, we will just remove the applywarp code path.

**Comment by coalsont** [comment]
For clarity, here is a capture of the current mask difference image (new workbench minus old applywarp), red positive, blue negative, black zero:

maskdiff

**Comment by glasserm** [comment]
I think the results are close enough.

File: fMRIVolume/scripts/OneStepResampling.sh (hunk 4 above)
Comment (mharms) [reply fMRIVolume/scripts/OneStepResampling.sh:138]: At least initially, I would suggest we leave the applywarp code path, but make `useWbResample=1` the default instead.

File: fMRIVolume/scripts/OneStepResampling.sh [hunk 7]
```diff
@@ -284,18 +284,19 @@ then
...
     
     #resample all-ones volume series with enclosing voxel to determine FOV coverage
     #yes, this is the entire length of the timeseries on purpose
-    fovcheck="$(mktemp --tmpdir OneStepResampleFovCheck_XXXXXX.nii.gz)"
-    tempfiles_add "$fovcheck"
+    fovcheck="$(tempfiles_create OneStepResampleFovCheck_XXXXXX.nii.gz)"
     wb_command -volume-math '1' "$fovcheck" -var x "$InputfMRI"
-    wb_command -volume-resample "$fovcheck" "$WD/$T1wImageFile.$FinalfMRIResolution.nii.gz" ENCLOSING_VOXEL "${OutputfMRI}_mask.nii.gz" "${xfmargs[@]}" -nifti-output-datatype INT32
+    #fsl's "nn" interpolation is just as pessimistic about the FoV as non-extrapolating interpolation
+    #so, in wb_command, use trilinear here instead of enclosing voxel to get the same FoV behavior as CUBIC
```
Comment (mharms) [inline fMRIVolume/scripts/OneStepResampling.sh:298]: Maybe add to the this comment: `(still not identical, but TRILINEAR closer than ENCLOSING_VOXEL)`
